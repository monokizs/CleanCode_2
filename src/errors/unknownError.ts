export class UnknownError extends Error {
    public constructor(message: string, public error?: Error) {
        super(message);
        this.name = this.constructor.name;
    }
}