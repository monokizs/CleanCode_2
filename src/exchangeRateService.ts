export class IExchangeRateService{

    public getExchangeRate(fromCurrency: string, toCurrency: string) {
        return 1;
    }

}