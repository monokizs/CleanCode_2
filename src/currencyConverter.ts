import { UnknownError } from "./errors/unknownError";
import { EndDateError } from "./errors/endDateError";
import { ServiceNotAvailableError } from "./errors/serviceNotAvailableError";

import { IExchangeRateService } from "./exchangeRateService"; 


export class CurrencyConverter {
    private readonly FIXED_AMOUNT = 100;
    
    constructor(private exchangeRateService: IExchangeRateService) { }
    
    public Convert(amount: number, fromCurrency: string, toCurrency: string): number{
        this.validateAmount(amount);
        const exchangeRate = this.getExchangeRate(fromCurrency, toCurrency); 
        //this.validateExchangeRate(exchangeRate);
        return amount * exchangeRate;
    }
    
    public GenerateConversionReport(fromCurrency: string, toCurrency: string, startDate: Date, endDate: Date): string {
        const conversions: number[] = [];
        const currentDate = new Date(startDate);
        
        if (endDate<startDate){
            throw new EndDateError('The enddate less than startdate.');
        } else {

            while (currentDate <= endDate) { 
                    let exchangeRate=0;
                    try {
                        exchangeRate = this.exchangeRateService.getExchangeRate(fromCurrency, toCurrency);
                    } catch (error) {
                        if (error instanceof ServiceNotAvailableError){
                            throw error;
                        }
                        throw new UnknownError('Unknown service error happened.', error as Error);
                    }
                    //this.validateExchangeRate(exchangeRate); 
                    this.calculateConversion(exchangeRate, conversions, currentDate);
            }
            
            return `Conversion Report:\n${conversions.join('\n')}`;
        }
    }
    
    private getExchangeRate(fromCurrency: string, toCurrency: string) {
        try {
            return this.exchangeRateService.getExchangeRate(fromCurrency, toCurrency);
        } catch (error) {
            if (error instanceof ServiceNotAvailableError){
                throw error;
            }
            
            throw new UnknownError('Unknown service error happened.', error as Error);
        }
        
    }
    
    private calculateConversion(exchangeRate: number, conversions: number[], currentDate: Date) {
        const convertedAmount = this.FIXED_AMOUNT * exchangeRate; // Assume a fixed amount for simplicity
        conversions.push(convertedAmount); 
        currentDate.setDate(currentDate.getDate() + 1);
    }
    
    /*private validateExchangeRate(exchangeRate: number) { 
        if (!exchangeRate) {
            throw new Error('Unable to fetch exchange rate.');
        }
    
        if (isNaN(exchangeRate)) {
            throw new Error('Invalid exchange rate.');
        }
    }*/
    
    private validateAmount(amount: number) { 
        if (isNaN(amount) || amount<0) {
            throw new Error('Invalid amount input.');
        }
    }

}
    