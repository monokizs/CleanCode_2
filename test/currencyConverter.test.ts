import { mock, mockReset } from "jest-mock-extended";
import { CurrencyConverter } from "../src/currencyConverter";
import { IExchangeRateService } from "../src/exchangeRateService";
import { MyError } from "../src/errors/myError";

/*jest.mock('../src/exchangeRateService', () => {
    return {
        IExchangeRateService: jest.fn().mockImplementation(() => {
            return {
                getExchangeRate: jest.fn().mockReturnValue(365.27)
            }
        })
    }
});*/

describe('Currency Converter test', () => {
    let sut: CurrencyConverter;
    const mockedIExchangeRateService = mock<IExchangeRateService>();

    beforeEach(()=>{
        mockReset(mockedIExchangeRateService);
        sut = new CurrencyConverter(mockedIExchangeRateService);
    })

    describe('Happy path', ()=> {
        it('should return the rate', () => {
            // Arrange
            mockedIExchangeRateService.getExchangeRate.mockReturnValue(365.27);    
            // Act
            const result = sut.Convert(100,"EUR","HUF");
    
            // Assert

            expect(mockedIExchangeRateService.getExchangeRate).toHaveBeenCalledTimes(1);
            expect(mockedIExchangeRateService.getExchangeRate).toHaveBeenCalledWith("EUR","HUF");
        })

        it('should return the conversion report tests', () => {
            // Arrange
            mockedIExchangeRateService.getExchangeRate.mockReturnValue(365.27); 

            // Act
            const result = sut.GenerateConversionReport("EUR","HUF", new Date('2024-04-01'), new Date('2024-04-04'));

            // Assert
            expect(result).toMatchSnapshot();
        })


    })
    
    describe('Error path', ()=> {
        it('should throw a Invalid amount input error if the amount does not a number', () => {
            // Arrange
            const errorMessage = "Invalid amount input.";
            const expectedError = new Error(errorMessage);
            mockedIExchangeRateService.getExchangeRate.mockImplementation(() => { throw expectedError });

            // Act
            expect(() => sut.Convert(100,"EUR","HUF")).toThrow(expectedError);
            expect(mockedIExchangeRateService.getExchangeRate).toHaveBeenCalledTimes(1);
            expect(mockedIExchangeRateService.getExchangeRate).toHaveBeenCalledWith("EUR","HUF");
        })

        it('should throw an Unable to fetch exchange rate error if the rate is empty', () => {
            // Arrange
            const errorMessage = "Unable to fetch exchange rate.";
            const expectedError = new Error(errorMessage);
            mockedIExchangeRateService.getExchangeRate.mockImplementation(() => { throw expectedError });

            // Act
            expect(() => sut.Convert(100,"GBP","HUF")).toThrow(expectedError);
            expect(mockedIExchangeRateService.getExchangeRate).toHaveBeenCalledTimes(1);
            expect(mockedIExchangeRateService.getExchangeRate).toHaveBeenCalledWith("GBP","HUF");
        })

        it('should throw an Exchange rate not found error if the rate is empty', () => {
            // Arrange
            const errorMessage = 'Rate error.';
            const expectedError = new Error(errorMessage);
            const myErrorMessage = 'Exchange rate not found.';
            const myExpectedError = new MyError(myErrorMessage, expectedError);
            mockedIExchangeRateService.getExchangeRate.mockImplementation(() => { throw expectedError });

            // Act
            expect(() => sut.Convert(100,"GBP","HUF")).toThrow(myExpectedError);
            expect(mockedIExchangeRateService.getExchangeRate).toHaveBeenCalledTimes(1);
            expect(mockedIExchangeRateService.getExchangeRate).toHaveBeenCalledWith("GBP","HUF");
            
        })


    })
})