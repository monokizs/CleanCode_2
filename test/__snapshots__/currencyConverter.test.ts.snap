// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Currency Converter test Happy path should return the conversion report tests 1`] = `
"Conversion Report:
36527
36527
36527
36527"
`;
